mm2src/coins/eth.rs

```diff
@@ -2670,150 +2679,155 @@ impl EthCoin {
         secret_hash: &[u8],
         amount: BigDecimal,
         expected_swap_contract_address: Address,
-    ) -> Box<dyn Future<Item = (), Error = String> + Send> {
-        let unsigned: UnverifiedTransaction = try_fus!(rlp::decode(payment_tx));
-        let tx = try_fus!(SignedEthTx::new(unsigned));
-        let sender = try_fus!(addr_from_raw_pubkey(sender_pub));
-        let expected_value = try_fus!(wei_from_big_decimal(&amount, self.decimals));
+    ) -> ValidatePaymentFut<()> {
+        let unsigned: UnverifiedTransaction = try_f!(rlp::decode(payment_tx));
+        let tx =
+            try_f!(SignedEthTx::new(unsigned).map_to_mm(|err| ValidatePaymentError::InvalidInput(err.to_string())));
+        let sender = try_f!(addr_from_raw_pubkey(sender_pub).map_to_mm(ValidatePaymentError::InvalidInput));
+        let expected_value = try_f!(wei_from_big_decimal(&amount, self.decimals));
         let selfi = self.clone();
         let secret_hash = secret_hash.to_vec();
         let fut = async move {
             let swap_id = selfi.etomic_swap_id(time_lock, &secret_hash);
-            let status = try_s!(
-                selfi
-                    .payment_status(expected_swap_contract_address, Token::FixedBytes(swap_id.clone()))
-                    .compat()
-                    .await
-            );
+            let status = selfi
+                .payment_status(expected_swap_contract_address, Token::FixedBytes(swap_id.clone()))
+                .compat()
+                .await
+                .map_to_mm(|e| ValidatePaymentError::Web3RpcError(Web3RpcError::Internal(e)))?;
```

Comment:
As we discussed with @shamardy, it's not clear if we convert `UtxoRpcError::ResponseParseError` or `UtxoRpcError::InvalidResponse` into `ValidatePaymentError::Transport`. Transport means an actual transport layer error.
Please add one more error variant to the `ValidatePaymentError`:
```rust
pub enum ValidatePaymentError {
  ...
  InvalidRpcResponse(String),
  Transport(String),
  ...
}
```
`InvalidRpcResponse` is not considered to be parsed or handle specially on the `validate_payment` function usage, but it will let the function user know that the function fails due to an unknown RPC response, not a Transport error.